include/envoy/http/header_map.h

```diff
@@ -348,6 +348,67 @@ class HeaderMap {
   virtual void addCopy(const LowerCaseString& key, const std::string& value) PURE;
 
   /**
+   * Set a reference header in the map. Both key and value MUST point to data that will live beyond
+   * the lifetime of any request/response using the string (since a codec may optimize for zero
+   * copy). Nothing will be copied.
+   *
+   * Calling setReference multiple times for the same header will result in only the last header
+   * being present in the HeaderMap.
+   *
+   * @param key specifies the name of the header to set; it WILL NOT be copied.
+   * @param value specifies the value of the header to set; it WILL NOT be copied.
+   */
+  virtual void setReference(const LowerCaseString& key, const std::string& value) PURE;
```

Comment:
It might be less verbose to just make this a bool arg on the existing `add*`, but I don't think it's a big deal.